### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by taxe10**
Companion PRs:
tomo: https://github.com/mlexchange/mlex_tomo_framework/pull/16
prefect worker: https://github.com/mlexchange/mlex_prefect_worker/pull/26
mlex_utils: https://github.com/mlexchange/mlex_utils/pull/5
highres_seg: https://github.com/mlexchange/mlex_highres_segmentation/pull/201

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1211868935400581

### CODE REVIEW

File: mlex_dlsia/main.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+from mlex_dlsia.utils.tiled import prepare_tiled_containers
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("yaml_path", type=str, help="path of yaml file for parameters")
+    parser.add_argument("--train", action="store_true", help="train the model")
```
Comment (xiaoyachong) [inline mlex_dlsia/main.py:25]: Shall we add `--inference` flag as well?

File: mlex_dlsia/network.py [hunk 2]
```diff
@@ -1,12 +1,122 @@
...
+    Input:
+        network: str, name of the DLSIA network to be loaded.
+        model_dir: str, path of the saved network.
+    Output:
+        net: loaded pre-trained network
+    """
+    if not os.path.isdir(model_dir):
+        raise ValueError(f"Model directory {model_dir} does not exist.")
+
+    if network_name == "DLSIA SMSNetEnsemble":
```
Comment (xiaoyachong) [inline mlex_dlsia/network.py]: I noticed that the segmentation app doesn’t support selecting `DLSIA SMSNetEnsemble`.

File: mlex_dlsia/train.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        trainer = Trainer(
+            net,
+            train_loader,
+            val_loader,
+            model_parameters.num_epochs,
+            criterion,
+            optimizer,
+            device,
+            dvclive=dvclive,
+            savepath=model_dir,
```
Comment (xiaoyachong) [inline mlex_dlsia/train.py]: Not directly related to this PR, but here’s a potential issue:
If we set `savepath=None` (meaning we don’t want to save the model or loss files because `MLflow` will handle it), the code may crash. There’s a potential issue in the `dlsia.core.train_scripts` package—`Trainer()` does not handle `savepath=None` correctly.

File: pyproject.toml [hunk 4]
```diff
@@ -1,7 +1,38 @@
...
+description = "DLSIA models for segmentation in MLExchange"
+readme = { file = "README.md", content-type = "text/markdown" }
+requires-python = ">=3.11"
+
+dependencies = [
+  "qlty==0.2.0",
+  "dvclive==3.44.0",
+  "dlsia==0.3.4",
+  "pydantic>=1.10.15",
+  "tiled[client]==0.2.0",
```
Comment (xiaoyachong) [inline pyproject.toml:27]: `mlflow==2.22.0` needs to be added here now that `requirements.txt` has been removed.

File: _tests/example_bad_params.yaml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  data_tiled_uri: 1
+  data_tiled_api_key: 2
+  mask_tiled_uri: 3
+  mask_tiled_api_key: 4
+  seg_tiled_uri: 5
+  seg_tiled_api_key: 6
+  uid_save: 7
+  uid_retrieve: 8
+  models_dir: .
+  job_name: bad_params_test
```
Comment (xiaoyachong) [inline _tests/example_bad_params.yaml:14]: In the current main branch, four new arguments — `mlflow_uri`, `mlflow_tracking_username`, `mlflow_tracking_password` and `mlflow_model` — have been added to `io_parameters`. This means that all `.yaml` files will need to be updated once the changes are merged into main.

Looking ahead, to align with the upcoming Prefect worker upgrade, the parameters `data_tiled_api_key`, `mask_tiled_api_key`, and `seg_tiled_api_key` will need to be removed. Additionally, `models_dir` will no longer be set (it will default to `None`), since models and DVC results will be stored in MLflow rather than in models_dir.  `mlflow_tracking_username` and `mlflow_tracking_password` will need to be removed, too.

File: _tests/test_3-train.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+def test_train_network(validated_params_with_raw):
+    io_parameters, model_parameters, raw_parameters = validated_params_with_raw
+    print(f"validated model parameters: {model_parameters}")
+    train_loader, val_loader = _get_mock_data(model_parameters)
+    item, mask = next(iter(train_loader))
+    try:
+        net = build_network(
+            model_parameters.network,
+            in_channels=1,
+            image_shape=(model_parameters.qlty_window, model_parameters.qlty_window),
```
Comment (xiaoyachong) [inline _tests/test_3-train.py:38]: `image_shape` does seem to resolve the previous bug with `TUNET3+`.

**Review by xiaoyachong — COMMENTED** [review]
Thanks for your PR! It looks good. The only potential issue is that it currently has conflicts with the main branch. Could you merge main into your branch and resolve the conflicts on your side? Once that’s done, I can test it together with the other PRs.

File: mlex_dlsia/network.py (hunk 2 above)
Comment (taxe10) [reply mlex_dlsia/network.py]: That is correct, it was at an "experimental" stage when they were added to this repo. We should test them and add them to the model options in the segmentation app.

**Comment by taxe10** [comment]
Companion PR: https://github.com/mlexchange/mlex_highres_segmentation/pull/201

File: mlex_dlsia/train.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+            # Clear out unnecessary variables from device memory
+            torch.cuda.empty_cache()
+        logger.info(f"{network_name} trained successfully.")
+
+        if model_parameters.network == "DLSIA SMSNetEnsemble":
+            net = baggin_smsnet_ensemble(networks=trained_nets)
+        else:
+            net = trained_nets[0]
+
+        # Log model to MLflow
```
Comment (xiaoyachong) [inline mlex_dlsia/train.py]: Lines 133–142 should be moved after line 121 and aligned with line 121 using the same indentation.
The reason is `log_model` and `log_artifacts` should be placed inside the `for idx, net in enumerate(networks)`: loop.

File: mlex_dlsia/parameters.py [hunk 8]
```diff
@@ -27,7 +27,6 @@ class IOParameters(BaseModel):
     models_dir: Optional[str] = Field(
         default=None, description="directory to save model results"
     )
-    # NEW: MLflow parameters
     mlflow_uri: Optional[str] = Field(
         default="http://mlflow:5000", description="MLflow tracking uri (optional)"
     )
```
Comment (xiaoyachong) [inline mlex_dlsia/parameters.py:32]: `mlflow_model` needs to be added, as it is used for inference.
```
  mlflow_model: Optional[str] = Field(
      default=None, description="MLflow model name (optional)"
  )
```

File: mlex_dlsia/network.py [hunk 9]
```diff
@@ -194,96 +280,19 @@ def build_smsnet_ensemble(
...
-def build_network(
-    network,
-    data_shape,
-    num_classes,
-    parameters,
-):
-    if len(data_shape) == 3:
-        in_channels = 1
-        image_shape = data_shape[1:]
+def baggin_smsnet_ensemble(networks=None, network_dir=None):
```
Comment (xiaoyachong) [inline mlex_dlsia/network.py]: `baggin_smsnet_ensemble` should load model from `MLflow`. It should be something like below (but I have not tested it):
```
def baggin_smsnet_ensemble(model_name):
        print(f"Loading ensemble models from MLflow registry: {model_name}")
        # Get all versions of the registered model
        client = mlflow.MlflowClient()
        model_versions = client.search_model_versions(f"name='{model_name}'")

        # Load all models
        list_of_models = []
        for mv in sorted(model_versions, key=lambda x: int(x.version)):
            model_uri = f"models:/{model_name}/{mv.version}"
            print(f"Loading model version {mv.version} from {model_uri}")
            model = mlflow.pytorch.load_model(model_uri)
            list_of_models.append(model)

        # Create ensemble using baggin
        net = model_baggin(models=list_of_models, model_type="classification")
        print(
            f"Ensemble created with {len(list_of_models)} models from MLflow registry"
        )
        return net
```

File: mlex_dlsia/main.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+            device,
+            model_dir,
+        )
+        logging.info("Training completed successfully.")
+    else:
+        if hasattr(io_parameters, "mlflow_model") and io_parameters.mlflow_model:
+            model_name = io_parameters.mlflow_model
+        else:
+            model_name = io_parameters.uid_retrieve
+        net = load_network(model_name)
```
Comment (xiaoyachong) [inline mlex_dlsia/main.py]: We may need to check `if network == "DLSIA SMSNetEnsemble"` to decide whether to call `load_network()` or `bagging_smsnet_ensemble()`.

`"DLSIA SMSNetEnsemble"` appears to be a more complex case, and I haven’t tested it with the application yet.

File: mlex_dlsia/train.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+            )
+            net, _ = trainer.train_segmentation()  # training happens here
+
+            trained_nets.append(net)
+            # Clear out unnecessary variables from device memory
+            torch.cuda.empty_cache()
+        logger.info(f"{network_name} trained successfully.")
+
+        if model_parameters.network == "DLSIA SMSNetEnsemble":
+            net = baggin_smsnet_ensemble(networks=trained_nets)
```
Comment (xiaoyachong) [inline mlex_dlsia/train.py]: Line 129 may have a bug, shall we call `model_baggin()` instead of `baggin_smsnet_ensemble()`.

**Review by xiaoyachong — COMMENTED** [review]
Thank you for your PR! I found a few issues with the refactor and left some comments.

**Review by xiaoyachong — APPROVED** [review]
@taxe10 @Wiebke This PR has been tested on my end alongside the companion PRs, and everything works as expected.
